cmd/gitserver/server/vcs_syncer_go_modules.go

```diff
@@ -0,0 +1,322 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/inconshreveable/log15"
+	"golang.org/x/mod/module"
+	modzip "golang.org/x/mod/zip"
+
+	dependenciesStore "github.com/sourcegraph/sourcegraph/internal/codeintel/dependencies/store"
+	"github.com/sourcegraph/sourcegraph/internal/conf/reposource"
+	"github.com/sourcegraph/sourcegraph/internal/errcode"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/gomodproxy"
+	"github.com/sourcegraph/sourcegraph/internal/repos"
+	"github.com/sourcegraph/sourcegraph/internal/unpack"
+	"github.com/sourcegraph/sourcegraph/internal/vcs"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+var placeholderGoDependency = func() *reposource.GoDependency {
+	dep, err := reposource.ParseGoDependency("sourcegraph.com/placeholder@v0.0.0")
+	if err != nil {
+		panic(fmt.Sprintf("expected placeholder dependency to parse but got %v", err))
+	}
+	return dep
+}()
+
+type GoModulesSyncer struct {
+	// Configuration object describing the connection to the npm registry.
+	connection *schema.GoModulesConnection
+	depsStore  repos.DependenciesStore
+	client     *gomodproxy.Client
```

Comment:
```suggestion
	connection schema.GoModulesConnection
	depsStore  repos.DependenciesStore
	client     gomodproxy.Client
```

AFAICT, the `connection` field and `client` field are always non-nil.

Reply:
The reason for passing pointers in isn't only to communicate the nullability of something but also for efficiency. No need to copy whole structs if we already have a pointer at hand, which is common practice to pass around in Go.

Reply:
Yes, it isn't a frequent operation, so wouldn't have a large impact, but again, it's not too common to pass large structs by value in Go. See all other Source constructors.

There are some related useful comments in the Go review comments wiki page:

- https://github.com/golang/go/wiki/CodeReviewComments#copying
- https://github.com/golang/go/wiki/CodeReviewComments#pass-values

Reply:
I think we have an over-use of pointers in our code base. Pointers should be used for the following reasons (in order, with exponentially decreasing favoritism):

- because the value should have reference semantics, not value semantics
- because the value may be nullable (strongly consider a useful zero-value here instead unless a primitive value)
- to reduce copies (only in critically hot paths identified by instrumenting an already correct solution)

I think the pointers here actually obscure the _meaning_ of the value as it's intended to be used. And since this isn't in a hot path affected by the pointers I'd say that we're only making the code less readable for no runtime benefit.

Reply:
I understand where you're coming from on the pointers usage and agree with that stack ranking, but I think we don't yet agree on copies being only worth reducing on critically hot-paths. That's probably a worthy discussion for Backend Crew, since we don't really have that codified in internal Go review guidelines.

Personally, I don't feel the pain from the use of pointers you mention, but I can see a case for adding some overhead for some understandability and guarantees in some places.

Reply:
> So why are we using pointers in this case? What is the benefit here over using value types? Seems like https://github.com/sourcegraph/sourcegraph/pull/32999#discussion_r837710108 for this PR?

Performance is always a consideration for me, and that link you shared was an explicit consideration about performance. Here it boils down to personal preference and habit? I have been writing Go [this way](https://github.com/golang/go/wiki/CodeReviewComments) for a long time, and obviously there are many others at Sourcegraph that would make the same choice.

So if you want to make a proposal for how we do this particular thing at large in this company, I'd welcome the discussion, but this doesn't seem like a big enough thing to riff over in this PR, and it's not obvious to me that it's a clear benefit.

Reply:
I see [Pass values](https://github.com/golang/go/wiki/CodeReviewComments#pass-values) section make a disclaimer that it _does not_ apply to large structs or structs that can grow. Is there a specific place in this document where it has _non-negative_ advice on using pointers?

Effective go also only seems to make the distinction between pointer and value types for _receivers_. I believe I've seen this advice before (pass larger structs by reference to save the copy) but can't point to where.

Reply:
> Is there a specific place in this document where it has non-negative advice on using pointers?

Not that I'm aware. That's the place. I think the logic behind that advice is that structs can and usually grow over time, and with passing pointers around the cost stays linear, where with values that get copied it's more than linear. The places where things get copied can also grow non-linearly if we don't pay attention to it — again, not a problem with pointers.